Modules/_abc.c

```diff
@@ -0,0 +1,617 @@
+        return NULL;
+    }
+    subclass = _PyObject_GetAttrId(instance, &PyId___class__);
+    /* Inline the cache checking. */
+    incache = _in_weak_set(self->abc_cache, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (incache > 0) {
+        Py_RETURN_TRUE;
+    }
+    subtype = (PyObject *)Py_TYPE(instance);
+    if (subtype == subclass) {
+        incache = _in_weak_set(self->abc_negative_cache, subclass);
+        if (incache < 0) {
+            return NULL;
+        }
+        if (self->abc_negative_cache_version == abc_invalidation_counter && incache) {
+            Py_RETURN_FALSE;
+        }
+        /* Fall back to the subclass check. */
+        return PyObject_CallMethod((PyObject *)self, "__subclasscheck__", "O", subclass);
+    }
+    result = PyObject_CallMethod((PyObject *)self, "__subclasscheck__", "O", subclass);
+    if (!result) {
+        return NULL;
+    }
+    if (result == Py_True) {
+        return Py_True;
+    }
+    Py_DECREF(result);
+    return PyObject_CallMethod((PyObject *)self, "__subclasscheck__", "O", subtype);
+}
+
+static PyObject *
+abcmeta_subclasscheck(abc *self, PyObject *args)
+{
+    PyObject *subclasses, *subclass = NULL;
+    PyObject *ok, *mro, *key, *rkey;
+    Py_ssize_t pos;
+    int incache, result;
+    if (!PyArg_UnpackTuple(args, "__subclasscheck__", 1, 1, &subclass)) {
+        return NULL;
+    }
+    /* TODO: clear the registry from dead refs from time to time
+       on iteration here (have a counter for this) or maybe during a .register() call */
+    /* TODO: Reset caches every n-th succes/failure correspondingly
+       so that they don't grow too large */
+
+    /* 1. Check cache. */
+    incache = _in_weak_set(self->abc_cache, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (incache > 0) {
+        Py_RETURN_TRUE;
+    }
+    /* 2. Check negative cache; may have to invalidate. */
+    incache = _in_weak_set(self->abc_negative_cache, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (self->abc_negative_cache_version < abc_invalidation_counter) {
+        /* Invalidate the negative cache. */
+        self->abc_negative_cache = PySet_New(NULL);
+        if (!self->abc_negative_cache) {
+            return NULL;
+        }
+        self->abc_negative_cache_version = abc_invalidation_counter;
+    } else if (incache) {
+        Py_RETURN_FALSE;
+    }
+    /* 3. Check the subclass hook. */
+    ok = PyObject_CallMethod((PyObject *)self, "__subclasshook__", "O", subclass);
+    if (!ok) {
+        return NULL;
+    }
+    if (ok == Py_True) {
+        if (!_add_weak_set(self->abc_cache, subclass)) {
+            Py_DECREF(ok);
+            return NULL;
+        }
+        return Py_True;
+    }
+    if (ok == Py_False) {
+        if (!_add_weak_set(self->abc_negative_cache, subclass)) {
+            Py_DECREF(ok);
+            return NULL;
+        }
+        return Py_False;
+    }
+    Py_DECREF(ok);
```

Comment:
This causes a change in behavior when a  `__subclasshook__` method returns something that is neither `NotImplemented` nor a `bool`. Previously, it raised an `AssertionError`, now it is silently ignored